src/transform/deadCode.ml

```diff
@@ -0,0 +1,68 @@
+open Batteries
+open GoblintCil
+
+
+(** Filter statements out of a block (recursively). CFG fields (prev, next, etc.) are no longer valid after calling.
+    Returns true if anything is left in block, false if the block is now empty.
+    Invariants:
+    - f (goto label) ==> f (labelled stmt), i.e. if a goto statement is not filtered out, the target may not be filtered out either.
+    - block may not contain switch statements *)
+let filter_map_block f (block : Cil.block) : bool =
+  (* blocks and statements: modify in place, then return true if should be kept *)
+  let rec impl_block block =
+    block.bstmts <- List.filter impl_stmt block.bstmts;
+    block.bstmts <> []
+  and impl_stmt stmt =
+    if not (f stmt) then false
+    else
+      match stmt.skind with
+      | If (_, b1, b2, _, _) ->
+        (* be careful to not short-circuit, since call to impl_block b2 is always needed for side-effects *)
+        let keep_b1, keep_b2 = impl_block b1, impl_block b2 in keep_b1 || keep_b2
+      | Switch _ -> failwith "switch statements must be removed"
+        (* handling switch statements correctly would be very difficult; consider that the switch
+        labels may be located within arbitrarily nested statements within the switch statement's block *)
+      | Loop (b, _, _, _, _) ->
+        impl_block b
+      | Block b ->
+        impl_block b
+      | sk -> true
+  in
+  impl_block block
+
+
+module RemoveDeadCode : Transform.S = struct
+  let transform (q : Transform.queries) (file : file) : unit =
+
+    (* whether a statement (might) still be live, and should therefore be kept *)
+    let stmt_live : stmt -> bool = not % q.must_be_dead in
+
+    (* whether a global function (might) still be live, and should therefore be kept *)
+    let fundec_live : fundec -> bool = not % q.must_be_uncalled in
+
+    (* step 1: remove statements found to be dead *)
+    Cil.iterGlobals file
+      (function
+      | GFun (fd, _) ->
+        (* invariants of filter_map_block satisfied: switch statements removed by CFG transform,
+           and a live label implies its target is live *)
+        filter_map_block stmt_live fd.sbody |> ignore
+      | _ -> ());
+
+    (* step 2: remove globals that are (transitively) unreferenced by live functions
+       - dead functions: removed by answering 'false' for isRoot
+       - unreferenced globals and types: removeUnusedTemps keeps only globals referenced by live functions *)
+    let open GoblintCil.Rmtmps in
+    let keepUnused0 = !keepUnused in
+    Fun.protect ~finally:(fun () -> keepUnused := keepUnused0) (fun () ->
+      keepUnused := false;
+      removeUnusedTemps ~isRoot:(function GFun (fd, _) -> fundec_live fd | _ -> false) file
```

Comment:
Wouldn't it be enough for just the `main` function to be root?

Reply:
Actually, isn't this even biased the incorrect way? If `main` has takes a function pointer to an uncalled function `foo`, then even considering just `main` as root would still mean that `foo` is found to be reachable from the root even when `isRoot` for it returns false, no? Since `Rmtmps` is entirely syntactical, it feels like it wouldn't really be able to remove much at all because it is a conservative syntactic thing, which, to remain on the safe side, cannot know to remove things which our analysis knows more precisely about.
Or am I mistaken about what `Rmtmps` does?

Reply:
No, your understanding of `Rmtmps` is correct; it is a purely syntactical transformation. The interesting part, removing code that Goblint finds to be semantically dead, happens in the first phase by filtering out statements within a function.

For the purposes of this transformation, I think an uncalled function whose address is nevertheless taken still counts as a usage of that function, and thus it should be retained. I'm also not sure what the reference would be replaced with, as once the address has been stored somewhere, it may be used in non-trivial ways.

As for just having `main` as the root: as long as there is a (chain of) syntactic references from `main` to any live function, using just `main` is enough. Since the analyzed program is considered complete, we don't need to worry about keeping any visible library functions. The CIL docs on `Rmtmps` mention functions with the "constructor" or "destructor" attribute, not sure if those are relevant to Goblint. And finally the CIL global initializer should be called by `main` automatically, as far as I understand. So in short: this is the way we decide which functions are dead for warnings, so I'm more confident keeping it this way. But if there's no edge case I've missed, I'll change this to compare to `mainfun`.

Reply:
> For the purposes of this transformation, I think an uncalled function whose address is nevertheless taken still counts as a usage of that function, and thus it should be retained. I'm also not sure what the reference would be replaced with, as once the address has been stored somewhere, it may be used in non-trivial ways.

That's true. Uncalled functions which have their address taken transforming into an empty function seems sensible.

> Since the analyzed program is considered complete, we don't need to worry about keeping any visible library functions. The CIL docs on `Rmtmps` mention functions with the "constructor" or "destructor" attribute, not sure if those are relevant to Goblint.

Goblint has its own `mainfun`, etc stuff similar to that, but the whole Linux kernel auto-harness isn't in a good state, so I think it's fine as is for now.